<div style="display: flex; background-color: #3F579F;">
    <h1 style="margin: auto; font-weight: bold; padding: 30px 30px 0px 30px;" align="center">Consumption needs of buildings - Project 4</h1>
</div>
<div style="display: flex; background-color: #3F579F; margin: auto; padding: 5px 30px 0px 30px;" >
    <h2 style="width: 100%; text-align: center; float: left;" align="center">| Analysis notebook |</h2>
</div>
<div style="display: flex; background-color: #3F579F; margin: auto; padding: 10px 30px 30px 30px;">
    <h3 style="width: 100%; text-align: center; font-size:26px; float: left;" align="center">Data Scientist course - OpenClassrooms</h3>
</div>

<div style="background-color: #506AB9;" >
    <h3 style="margin: auto; padding: 20px; color:#fff; ">1. Libraries and functions</h3>
</div>

<div style="background-color: #6D83C5;" >
    <h4 style="margin: auto; padding: 20px; color:#fff; ">1.1. Libraries and functions</h3>
</div>

In [1]:
import os
import io
import gc
import ast
from math import prod
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec
import seaborn as sns

%matplotlib inline

sns.set_theme(style="darkgrid")

<div style="background-color: #6D83C5;" >
    <h4 style="margin: auto; padding: 20px; color:#fff; ">1.2. Functions declaration</h4>
</div>

In [57]:
def df_analysis(df, name_df, *args, **kwargs):
    """
    Method used for analyzing on the DataFrame.

    Parameters:
    -----------------
        df (pandas.DataFrame): Dataset to analyze
        name_df (str): Dataset name
        
        *args, **kwargs:
        -----------------
            columns (list): Dataframe keys in list format
            flag (str): Flag to show complete information about the dataset to analyse
                        "complete" shows all information about the dataset

    Returns:
    -----------------
        None. 
        Print the analysis on the Dataset. 
    """
    
    # Getting the variables
    columns = kwargs.get("columns", None)
    type_analysis = kwargs.get("type_analysis", None)
    
    ORDERING_COMPLETE = [
        "name", "type", "records", "unique", "# NaN", "% NaN", "mean", "min", "25%", "50%", "75%", "max", "std"
    ]
    
    # Calculating the memory usage based on dataframe.info()
    buf = io.StringIO()
    df.info(buf=buf)
    memory_usage = buf.getvalue().split('\n')[-2]
    
    if df.empty:
        print("The", name_df, "dataset is empty. Please verify the file.")
    else:
        empty_cols = [col for col in df.columns if df[col].isna().all()] # identifying empty columns
        df_rows_duplicates = df[df.duplicated()] #identifying full duplicates rows
        
        # Creating a dataset based on Type object and records by columns
        type_cols = df.dtypes.apply(lambda x: x.name).to_dict() 
        df_resume = pd.DataFrame(list(type_cols.items()), columns = ["name", "type"])
        df_resume["records"] = list(df.count())
        df_resume["# NaN"] = list(df.isnull().sum())
        df_resume["% NaN"] = list(((df.isnull().sum() / len(df.index))*100).round(2))
        
        print("\nAnalysis of", name_df, "dataset")
        print("--------------------------------------------------------------------")
        print("- Dataset shape:                 ", df.shape[0], "rows and", df.shape[1], "columns")
        print("- Total of NaN values:           ", df.isna().sum().sum())
        print("- Percentage of NaN:             ", round((df.isna().sum().sum() / prod(df.shape)) * 100, 2), "%")
        print("- Total of full duplicates rows: ", df_rows_duplicates.shape[0])
        print("- Total of empty rows:           ", df.shape[0] - df.dropna(axis="rows", how="all").shape[0]) if df.dropna(axis="rows", how="all").shape[0] < df.shape[0] else \
                    print("- Total of empty rows:            0")
        print("- Total of empty columns:        ", len(empty_cols))
        print("  + The empty column is:         ", empty_cols) if len(empty_cols) == 1 else \
                    print("  + The empty column are:         ", empty_cols) if len(empty_cols) >= 1 else None
        print("- Unique indexes:                ", df.index.is_unique)
        
        if columns is not None:
            print("\n- The key(s):", columns, "is not present multiple times in the dataframe.\n  It CAN be used as a primary key.") if df.size == df.drop_duplicates(columns).size else \
                print("\n- The key(s):", columns, "is present multiple times in the dataframe.\n  It CANNOT be used as a primary key.")
        
        pd.set_option("display.max_rows", None) # show full of showing rows
        pd.set_option("display.max_columns", None) # show full of showing cols
        pd.set_option("max_colwidth", None) # show full width of showing cols
        
        if type_analysis is None or type_analysis != "complete":
            print("\n- Type object and records by columns      (",memory_usage,")")
            print("--------------------------------------------------------------------")
        elif type_analysis == "complete":
            df_resume["unique"] = list(df.nunique())
            df_desc = pd.DataFrame(df.describe().T).reset_index()
            df_desc = df_desc.rename(columns={"index": "name"})
            df_resume = df_resume.merge(right=df_desc[["name", "mean", "min", "25%", "50%", "75%", "max", "std"]], on="name", how="left")
            df_resume = df_resume[ORDERING_COMPLETE]
            print("\n- Type object and records by columns      (",memory_usage,")")
            print("--------------------------------------------------------------------")
        
        display(df_resume.sort_values("records", ascending=False))
        
        pd.reset_option("display.max_rows") # reset max of showing rows
        pd.reset_option("display.max_columns") # reset max of showing cols
        pd.reset_option("display.max_colwidth") # reset width of showing cols
        
        # deleting dataframe to free memory
        if type_analysis == "complete":
            del [[df_resume, df_desc]]
            gc.collect()
            df_resume, df_desc = (pd.DataFrame() for i in range(2))
        else:
            del df_resume
            gc.collect()
            df_resume = pd.DataFrame()

<div style="background-color: #506AB9;" >
    <h3 style="margin: auto; padding: 20px; color:#fff; ">2. Importing files</h3>
</div>

<div style="background-color: #6D83C5;" >
    <h4 style="margin: auto; padding: 20px; color:#fff; ">2.1. Importing and preparing files</h4>
</div>

In [3]:
data_2015 = pd.read_csv(r"datasets\2015-building-energy-benchmarking.csv")
data_2016 = pd.read_csv(r"datasets\2016-building-energy-benchmarking.csv")

<div style="background-color: #6D83C5;" >
    <h4 style="margin: auto; padding: 20px; color:#fff; ">2.2. Initial analysis</h4>
</div>

<div class="alert alert-block alert-info">
Making <b>the initial analysis</b>
</div>

In [4]:
df_analysis(data_2015, "data_2015")


Analysis of data_2015 dataset
--------------------------------------------------------------------
- Dataset shape:                  3340 rows and 47 columns
- Total of NaN values:            26512
- Percentage of NaN:              16.89 %
- Total of full duplicates rows:  0
- Total of empty rows:            0
- Total of empty columns:         0
- Unique indexes:                 True

- Type object and records by columns   ( memory usage: 1.2+ MB )
---------------------------------------------------------------------------


name     type  \
0                                                   OSEBuildingID    int64   
9                                                       YearBuilt    int64   
40                                               ComplianceStatus   object   
1                                                        DataYear    int64   
14                                         PropertyGFABuilding(s)    int64   
13                                             PropertyGFAParking    int64   
12                                               PropertyGFATotal    int64   
10                                              NumberofBuildings    int64   
46                                                      Zip Codes    int64   
8                                                    Neighborhood   object   
4                                                    PropertyName   object   
3                                             PrimaryPropertyType   object   
2                                                    BuildingType   object   
7                                             CouncilDistrictCode    int64   
6                                                        Location   object   
38                                                    DefaultData   object   
5                                   TaxParcelIdentificationNumber   object   
45                                                      SPD Beats  float64   
43  Seattle Police Department Micro Community Policing Plan Areas  float64   
11                                                 NumberofFloors  float64   
29                                          SiteEnergyUseWN(kBtu)  float64   
36                                   GHGEmissions(MetricTonsCO2e)  float64   
35                                             OtherFuelUse(kBtu)  float64   
34                                               NaturalGas(kBtu)  float64   
33                                             NaturalGas(therms)  float64   
32                                              Electricity(kBtu)  float64   
31                                               Electricity(kWh)  float64   
30                                                 SteamUse(kBtu)  float64   
37                              GHGEmissionsIntensity(kgCO2e/ft2)  float64   
28                                            SiteEnergyUse(kBtu)  float64   
27                                           SourceEUIWN(kBtu/sf)  float64   
26                                             SourceEUI(kBtu/sf)  float64   
25                                             SiteEUIWN(kBtu/sf)  float64   
24                                               SiteEUI(kBtu/sf)  float64   
15                                      ListOfAllPropertyUseTypes   object   
16                                         LargestPropertyUseType   object   
17                                      LargestPropertyUseTypeGFA  float64   
23                                                ENERGYSTARScore  float64   
19                                SecondLargestPropertyUseTypeGFA  float64   
18                                   SecondLargestPropertyUseType   object   
20                                    ThirdLargestPropertyUseType   object   
21                                 ThirdLargestPropertyUseTypeGFA  float64   
42                                             2010 Census Tracts  float64   
44                                         City Council Districts  float64   
22                                       YearsENERGYSTARCertified   object   
41                                                        Outlier   object   
39                                                        Comment   object   

    records  # NaN  % NaN  
0      3340      0   0.00  
9      3340      0   0.00  
40     3340      0   0.00  
1      3340      0   0.00  
14     3340      0   0.00  
13     3340      0   0.00  
12     3340      0   0.00  
10     3340      0   0.00  
46     3340      0   0.00  
8      3340      0   0.00  
4      3

In [5]:
pd.set_option("display.max_rows", None) # show full of showing rows
pd.set_option("display.max_columns", None) # show full of showing cols
pd.set_option("max_colwidth", None) # show full width of showing cols

In [6]:
data_2015.head(3)

OSEBuildingID  DataYear    BuildingType PrimaryPropertyType  \
0              1      2015  NonResidential               Hotel   
1              2      2015  NonResidential               Hotel   
2              3      2015  NonResidential               Hotel   

           PropertyName TaxParcelIdentificationNumber  \
0  MAYFLOWER PARK HOTEL                     659000030   
1       PARAMOUNT HOTEL                     659000220   
2          WESTIN HOTEL                     659000475   

                                                                                                                                                       Location  \
0  {'latitude': '47.61219025', 'longitude': '-122.33799744', 'human_address': '{"address": "405 OLIVE WAY", "city": "SEATTLE", "state": "WA", "zip": "98101"}'}   
1    {'latitude': '47.61310583', 'longitude': '-122.33335756', 'human_address': '{"address": "724 PINE ST", "city": "SEATTLE", "state": "WA", "zip": "98101"}'}   
2   {'latitude': '47.61334897', 'longitude': '-122.33769944', 'human_address': '{"address": "1900 5TH AVE", "city": "SEATTLE", "state": "WA", "zip": "98101"}'}   

   CouncilDistrictCode Neighborhood  YearBuilt  NumberofBuildings  \
0                    7     DOWNTOWN       1927                  1   
1                    7     DOWNTOWN       1996                  1   
2                    7     DOWNTOWN       1969                  1   

   NumberofFloors  PropertyGFATotal  PropertyGFAParking  \
0            12.0             88434                   0   
1            11.0            103566               15064   
2            41.0            961990                   0   

   PropertyGFABuilding(s)      ListOfAllPropertyUseTypes  \
0                   88434                          Hotel   
1                   88502     Hotel, Parking, Restaurant   
2                  961990  Hotel, Parking, Swimming Pool   

  LargestPropertyUseType  LargestPropertyUseTypeGFA  \
0                  Hotel                    88434.0   
1                  Hotel                    83880.0   
2                  Hotel                   757243.0   

  SecondLargestPropertyUseType  SecondLargestPropertyUseTypeGFA  \
0                          NaN                              NaN   
1                      Parking                          15064.0   
2                      Parking                         100000.0   

  ThirdLargestPropertyUseType  ThirdLargestPropertyUseTypeGFA  \
0                         NaN                             NaN   
1                  Restaurant                          4622.0   
2               Swimming Pool                             0.0   

  YearsENERGYSTARCertified  ENERGYSTARScore  SiteEUI(kBtu/sf)  \
0                      NaN             65.0              78.9   
1                      NaN             51.0              94.4   
2                      NaN             18.0              96.6   

   SiteEUIWN(kBtu/sf)  SourceEUI(kBtu/sf)  SourceEUIWN(kBtu/sf)  \
0                80.3               173.5                 175.1   
1                99.0               191.3                 195.2   
2                99.7               242.7                 246.5   

   SiteEnergyUse(kBtu)  SiteEnergyUseWN(kBtu)  SteamUse(kBtu)  \
0            6981428.0              7097539.0       2023032.0   
1            8354235.0              8765788.0             0.0   
2           73130656.0             75506272.0      19660404.0   

   Electricity(kWh)  Electricity(kBtu)  NaturalGas(therms)  NaturalGas(kBtu)  \
0         1080307.0          3686160.0             12724.0         1272388.0   
1         1144563.0          3905411.0             44490.0         4448985.0   
2        14583930.0         49762435.0             37099.0         3709900.0   

   OtherFuelUse(kBtu)  GHGEmissions(MetricTonsCO2e)  \
0                 0.0                        249.43   
1                 0.0                        263.51   
2                 0.0                       2061.48   

   GHGEmissionsIn

In [7]:
df_analysis(data_2016, "data_2016")


Analysis of data_2016 dataset
--------------------------------------------------------------------
- Dataset shape:                  3376 rows and 46 columns
- Total of NaN values:            19952
- Percentage of NaN:              12.85 %
- Total of full duplicates rows:  0
- Total of empty rows:            0
- Total of empty columns:         1
  + The empty column is:          ['Comments']
- Unique indexes:                 True

- Type object and records by columns   ( memory usage: 1.2+ MB )
---------------------------------------------------------------------------


name     type  records  # NaN   % NaN
0                     OSEBuildingID    int64     3376      0    0.00
11                     Neighborhood   object     3376      0    0.00
1                          DataYear    int64     3376      0    0.00
40                      DefaultData     bool     3376      0    0.00
42                 ComplianceStatus   object     3376      0    0.00
19           PropertyGFABuilding(s)    int64     3376      0    0.00
18               PropertyGFAParking    int64     3376      0    0.00
17                 PropertyGFATotal    int64     3376      0    0.00
14                        YearBuilt    int64     3376      0    0.00
13                        Longitude  float64     3376      0    0.00
12                         Latitude  float64     3376      0    0.00
16                   NumberofFloors    int64     3376      0    0.00
10              CouncilDistrictCode    int64     3376      0    0.00
4                      PropertyName   object     3376      0    0.00
2                      BuildingType   object     3376      0    0.00
9     TaxParcelIdentificationNumber   object     3376      0    0.00
3               PrimaryPropertyType   object     3376      0    0.00
7                             State   object     3376      0    0.00
6                              City   object     3376      0    0.00
5                           Address   object     3376      0    0.00
33              SiteEnergyUse(kBtu)  float64     3371      5    0.15
34            SiteEnergyUseWN(kBtu)  float64     3370      6    0.18
30               SiteEUIWN(kBtu/sf)  float64     3370      6    0.18
29                 SiteEUI(kBtu/sf)  float64     3369      7    0.21
15                NumberofBuildings  float64     3368      8    0.24
32             SourceEUIWN(kBtu/sf)  float64     3367      9    0.27
44                TotalGHGEmissions  float64     3367      9    0.27
38               NaturalGas(therms)  float64     3367      9    0.27
37                Electricity(kBtu)  float64     3367      9    0.27
36                 Electricity(kWh)  float64     3367      9    0.27
35                   SteamUse(kBtu)  float64     3367      9    0.27
39                 NaturalGas(kBtu)  float64     3367      9    0.27
45            GHGEmissionsIntensity  float64     3367      9    0.27
31               SourceEUI(kBtu/sf)  float64     3367      9    0.27
20        ListOfAllPropertyUseTypes   object     3367      9    0.27
8                           ZipCode  float64     3360     16    0.47
22        LargestPropertyUseTypeGFA  float64     3356     20    0.59
21           LargestPropertyUseType   object     3356     20    0.59
28                  ENERGYSTARScore  float64     2533    843   24.97
24  SecondLargestPropertyUseTypeGFA  float64     1679   1697   50.27
23     SecondLargestPropertyUseType   object     1679   1697   50.27
26   ThirdLargestPropertyUseTypeGFA  float64      596   2780   82.35
25      ThirdLargestPropertyUseType   object      596   2780   82.35
27         YearsENERGYSTARCertified   object      119   3257   96.48
43                          Outlier   object       32   3344   99.05
41                         Comments  float64        0   3376  100.00

In [8]:
pd.set_option("display.max_rows", None) # show full of showing rows
pd.set_option("display.max_columns", None) # show full of showing cols
pd.set_option("max_colwidth", None) # show full width of showing cols

In [9]:
data_2016.head(5)

OSEBuildingID  DataYear    BuildingType PrimaryPropertyType  \
0              1      2016  NonResidential               Hotel   
1              2      2016  NonResidential               Hotel   
2              3      2016  NonResidential               Hotel   
3              5      2016  NonResidential               Hotel   
4              8      2016  NonResidential               Hotel   

                  PropertyName          Address     City State  ZipCode  \
0         Mayflower park hotel    405 Olive way  Seattle    WA  98101.0   
1              Paramount Hotel  724 Pine street  Seattle    WA  98101.0   
2      5673-The Westin Seattle  1900 5th Avenue  Seattle    WA  98101.0   
3                    HOTEL MAX   620 STEWART ST  Seattle    WA  98101.0   
4  WARWICK SEATTLE HOTEL (ID8)    401 LENORA ST  Seattle    WA  98121.0   

  TaxParcelIdentificationNumber  CouncilDistrictCode Neighborhood  Latitude  \
0                    0659000030                    7     DOWNTOWN  47.61220   
1                    0659000220                    7     DOWNTOWN  47.61317   
2                    0659000475                    7     DOWNTOWN  47.61393   
3                    0659000640                    7     DOWNTOWN  47.61412   
4                    0659000970                    7     DOWNTOWN  47.61375   

   Longitude  YearBuilt  NumberofBuildings  NumberofFloors  PropertyGFATotal  \
0 -122.33799       1927                1.0              12             88434   
1 -122.33393       1996                1.0              11            103566   
2 -122.33810       1969                1.0              41            956110   
3 -122.33664       1926                1.0              10             61320   
4 -122.34047       1980                1.0              18            175580   

   PropertyGFAParking  PropertyGFABuilding(s)      ListOfAllPropertyUseTypes  \
0                   0                   88434                          Hotel   
1               15064                   88502     Hotel, Parking, Restaurant   
2              196718                  759392                          Hotel   
3                   0                   61320                          Hotel   
4               62000                  113580  Hotel, Parking, Swimming Pool   

  LargestPropertyUseType  LargestPropertyUseTypeGFA  \
0                  Hotel                    88434.0   
1                  Hotel                    83880.0   
2                  Hotel                   756493.0   
3                  Hotel                    61320.0   
4                  Hotel                   123445.0   

  SecondLargestPropertyUseType  SecondLargestPropertyUseTypeGFA  \
0                          NaN                              NaN   
1                      Parking                          15064.0   
2                          NaN                              NaN   
3                          NaN                              NaN   
4                      Parking                          68009.0   

  ThirdLargestPropertyUseType  ThirdLargestPropertyUseTypeGFA  \
0                         NaN                             NaN   
1                  Restaurant                          4622.0   
2                         NaN                             NaN   
3                         NaN                             NaN   
4               Swimming Pool                             0.0   

  YearsENERGYSTARCertified  ENERGYSTARScore  SiteEUI(kBtu/sf)  \
0                      NaN             60.0         81.699997   
1                      NaN             61.0         94.800003   
2                      NaN             43.0         96.000000   
3                      NaN             56.0        110.800003   
4                      NaN             75.0        114.800003   

   SiteEUIWN(kBtu/sf)  SourceEUI(kBtu/sf)  SourceEUIWN(kBtu/sf)  \
0           84.300003          182.500000            189.000000   
1           97.900002          176.100006            179.399994   
2

<div style="background-color: #506AB9;" >
    <h3 style="margin: auto; padding: 20px; color:#fff; ">3. Cleaning</h3>
</div>

<div style="background-color: #6D83C5;" >
    <h4 style="margin: auto; padding: 20px; color:#fff; ">3.1. Formatting columns</h4>
</div>

<div class="alert alert-block alert-info">
    Splitting the featured <b>Location</b> in <b>data_2015</b>
</div>

In [10]:
df_temp = data_2015["Location"].apply(lambda x: ast.literal_eval(x))
data_2015[["Latitude", "Longitude", "HumanAddress"]] = df_temp.apply(pd.Series)

In [11]:
data_2015.head(2)

OSEBuildingID  DataYear    BuildingType PrimaryPropertyType  \
0              1      2015  NonResidential               Hotel   
1              2      2015  NonResidential               Hotel   

           PropertyName TaxParcelIdentificationNumber  \
0  MAYFLOWER PARK HOTEL                     659000030   
1       PARAMOUNT HOTEL                     659000220   

                                                                                                                                                       Location  \
0  {'latitude': '47.61219025', 'longitude': '-122.33799744', 'human_address': '{"address": "405 OLIVE WAY", "city": "SEATTLE", "state": "WA", "zip": "98101"}'}   
1    {'latitude': '47.61310583', 'longitude': '-122.33335756', 'human_address': '{"address": "724 PINE ST", "city": "SEATTLE", "state": "WA", "zip": "98101"}'}   

   CouncilDistrictCode Neighborhood  YearBuilt  NumberofBuildings  \
0                    7     DOWNTOWN       1927                  1   
1                    7     DOWNTOWN       1996                  1   

   NumberofFloors  PropertyGFATotal  PropertyGFAParking  \
0            12.0             88434                   0   
1            11.0            103566               15064   

   PropertyGFABuilding(s)   ListOfAllPropertyUseTypes LargestPropertyUseType  \
0                   88434                       Hotel                  Hotel   
1                   88502  Hotel, Parking, Restaurant                  Hotel   

   LargestPropertyUseTypeGFA SecondLargestPropertyUseType  \
0                    88434.0                          NaN   
1                    83880.0                      Parking   

   SecondLargestPropertyUseTypeGFA ThirdLargestPropertyUseType  \
0                              NaN                         NaN   
1                          15064.0                  Restaurant   

   ThirdLargestPropertyUseTypeGFA YearsENERGYSTARCertified  ENERGYSTARScore  \
0                             NaN                      NaN             65.0   
1                          4622.0                      NaN             51.0   

   SiteEUI(kBtu/sf)  SiteEUIWN(kBtu/sf)  SourceEUI(kBtu/sf)  \
0              78.9                80.3               173.5   
1              94.4                99.0               191.3   

   SourceEUIWN(kBtu/sf)  SiteEnergyUse(kBtu)  SiteEnergyUseWN(kBtu)  \
0                 175.1            6981428.0              7097539.0   
1                 195.2            8354235.0              8765788.0   

   SteamUse(kBtu)  Electricity(kWh)  Electricity(kBtu)  NaturalGas(therms)  \
0       2023032.0         1080307.0          3686160.0             12724.0   
1             0.0         1144563.0          3905411.0             44490.0   

   NaturalGas(kBtu)  OtherFuelUse(kBtu)  GHGEmissions(MetricTonsCO2e)  \
0         1272388.0                 0.0                        249.43   
1         4448985.0                 0.0                        263.51   

   GHGEmissionsIntensity(kgCO2e/ft2) DefaultData Comment ComplianceStatus  \
0                               2.64          No     NaN        Compliant   
1                               2.38          No     NaN        Compliant   

  Outlier  2010 Census Tracts  \
0     NaN                 NaN   
1     NaN                 NaN   

   Seattle Police Department Micro Community Policing Plan Areas  \
0                                                           14.0   
1                                                           14.0   

   City Council Districts  SPD Beats  Zip Codes     Latitude      Longitude  \
0                     NaN       31.0      18081  47.61219025  -122.33799744   
1                     NaN       31.0      18081  47.61310583  -122.33335756   

                                                                     HumanAddress  
0  {"address": "405 OLIVE WAY", "city": "SEATTLE", "state": "WA", "zip": "98101"}  
1    {"address": "724 PINE ST", "city": "SEATTLE", "state": "WA", "zip": "98101"}

<div class="alert alert-block alert-info">
    Based on the <b>official site of Seattle Open Data</b>, we can see the dataset <b>metadata</b> to know all features<br>
    <b>https://data.seattle.gov/dataset/2015-Building-Energy-Benchmarking/h7rm-fz6m</b>
    <ul style="list-style-type: square;">
        <li><b>GHGEmissions(MetricTonsCO2e)</b> in <b>data_2015</b> is the same feature as <b>TotalGHGEmissions</b> in <b>data_2016</b></li>
        <li><b>GHGEmissionsIntensity(kgCO2e/ft2)</b> in <b>data_2015</b> is the same feature as <b>GHGEmissionsIntensity</b> in <b>data_2016</b></li>
    </ul>
    
     On the other hands, we have the following
    <ul style="list-style-type: square;">
        <li><b>Zip Codes</b> in <b>data_2015</b> is the same feature as <b>ZipCode</b> in <b>data_2016</b></li>
    </ul>
</div>

In [12]:
data_2015["ZipCode"] = data_2015["Zip Codes"]
data_2016["GHGEmissionsIntensity(kgCO2e/ft2)"] = data_2016["GHGEmissionsIntensity"]
data_2016["GHGEmissions(MetricTonsCO2e)"] = data_2016["TotalGHGEmissions"]

data_2015.drop(labels=["Zip Codes"], axis=1, inplace=True)
data_2016.drop(labels=["GHGEmissionsIntensity", "TotalGHGEmissions"], axis=1, inplace=True)

<div class="alert alert-block alert-info">
<b>Comparing</b> the match between columns <b>before</b> making the <b>concatenation</b>
</div>

In [13]:
LIST_COL, LIST_ORIGIN = ([] for i in range(2))

for col in data_2015.columns:
    if col not in list(data_2016.columns):
        LIST_COL.append(col)
        LIST_ORIGIN.append("data_2015")
        
for col in data_2016.columns:
    if col not in list(data_2015.columns):
        LIST_COL.append(col)
        LIST_ORIGIN.append("data_2016")

df_columns_not_matches = pd.DataFrame({
    "column": LIST_COL,
    "origin": LIST_ORIGIN
})

In [14]:
df_columns_not_matches.sort_values("column")

column     origin
3                                              2010 Census Tracts  data_2015
8                                                         Address  data_2016
9                                                            City  data_2016
5                                          City Council Districts  data_2015
2                                                         Comment  data_2015
11                                                       Comments  data_2016
7                                                    HumanAddress  data_2015
0                                                        Location  data_2015
1                                              OtherFuelUse(kBtu)  data_2015
6                                                       SPD Beats  data_2015
4   Seattle Police Department Micro Community Policing Plan Areas  data_2015
10                                                          State  data_2016

<div class="alert alert-block alert-info">
<b>Deleting</b> the columns that aren't matched between the two datasets
</div>

In [15]:
data_2015.drop(labels=list(df_columns_not_matches.loc[df_columns_not_matches["origin"]=="data_2015", "column"]), axis=1, inplace=True)
data_2016.drop(labels=list(df_columns_not_matches.loc[df_columns_not_matches["origin"]=="data_2016", "column"]), axis=1, inplace=True)

<div class="alert alert-block alert-info">
<b>Comparing</b> the types of each par of column <b>before</b> making the <b>concatenation</b>
    
</div>

In [16]:
data_2015 = data_2015.reindex(sorted(data_2015.columns), axis=1)
data_2016 = data_2016.reindex(sorted(data_2016.columns), axis=1)

In [17]:
def types_different(data_2015, data_2016):
    
    LIST_COL_DIFFERENT, LIST_DTYPE_2015, LIST_DTYPE_2016 = [[] for i in range(3)]
    
    for col_2015, col_2016 in zip(data_2015.columns, data_2016.columns):
        if data_2015[col_2015].dtype != data_2016[col_2016].dtype:
            LIST_COL_DIFFERENT.append(col_2015)
            LIST_DTYPE_2015.append(str(data_2015[col_2015].dtype))
            LIST_DTYPE_2016.append(str(data_2016[col_2016].dtype))
            
    df_types_different = pd.DataFrame({
                            "column": LIST_COL_DIFFERENT,
                            "2015": LIST_DTYPE_2015,
                            "2016": LIST_DTYPE_2016
                        })
    
    return df_types_different

In [18]:
df_types_different = types_different(data_2015, data_2016)
df_types_different

column     2015     2016
0        DefaultData   object     bool
1           Latitude   object  float64
2          Longitude   object  float64
3  NumberofBuildings    int64  float64
4     NumberofFloors  float64    int64
5            ZipCode    int64  float64

<div class="alert alert-block alert-info">
<b>Fixing column types datetime</b>
</div>

In [19]:
data_2015["DefaultData"] = data_2015["DefaultData"].astype("bool")
data_2015["Latitude"] = data_2015["Latitude"].astype("float64")
data_2015["Longitude"] = data_2015["Longitude"].astype("float64")

data_2016["NumberofBuildings"] = data_2016["NumberofBuildings"].fillna(0)
data_2016["NumberofBuildings"] = data_2016["NumberofBuildings"].astype("int64")

data_2015["NumberofFloors"] = data_2015["NumberofFloors"].fillna(0)
data_2015["NumberofFloors"] = data_2015["NumberofFloors"].astype("int64")

data_2016["ZipCode"] = data_2016["ZipCode"].fillna(0)
data_2016["ZipCode"] = data_2016["ZipCode"].astype("int64")

df_types_different = types_different(data_2015, data_2016)
df_types_different

Empty DataFrame
Columns: [column, 2015, 2016]
Index: []

<div class="alert alert-block alert-info">
    <b>Deleting</b> the columns over <b>80%</b> of <b>missing-values</b>
</div>

In [20]:
data_2015 = data_2015.loc[:, data_2015.isnull().mean() < .8]
data_2016 = data_2016.loc[:, data_2016.isnull().mean() < .8]

<div class="alert alert-block alert-info">
    <b>Concatenating</b> the datasets
</div>

In [21]:
data = pd.concat([data_2015, data_2016], ignore_index=True)

In [22]:
df_analysis(data, "data")


Analysis of data dataset
--------------------------------------------------------------------
- Dataset shape:                  6716 rows and 38 columns
- Total of NaN values:            9264
- Percentage of NaN:              3.63 %
- Total of full duplicates rows:  0
- Total of empty rows:            0
- Total of empty columns:         0
- Unique indexes:                 True

- Type object and records by columns   ( memory usage: 1.9+ MB )
---------------------------------------------------------------------------


name     type  records  # NaN  % NaN
0                        BuildingType   object     6716      0   0.00
14                          Longitude  float64     6716      0   0.00
36                          YearBuilt    int64     6716      0   0.00
25                       PropertyName   object     6716      0   0.00
24                   PropertyGFATotal    int64     6716      0   0.00
23                 PropertyGFAParking    int64     6716      0   0.00
22             PropertyGFABuilding(s)    int64     6716      0   0.00
21                PrimaryPropertyType   object     6716      0   0.00
20                      OSEBuildingID    int64     6716      0   0.00
1                    ComplianceStatus   object     6716      0   0.00
18                  NumberofBuildings    int64     6716      0   0.00
17                       Neighborhood   object     6716      0   0.00
19                     NumberofFloors    int64     6716      0   0.00
3                            DataYear    int64     6716      0   0.00
4                         DefaultData     bool     6716      0   0.00
2                 CouncilDistrictCode    int64     6716      0   0.00
12                           Latitude  float64     6716      0   0.00
37                            ZipCode    int64     6716      0   0.00
35      TaxParcelIdentificationNumber   object     6714      2   0.03
30                SiteEnergyUse(kBtu)  float64     6701     15   0.22
29                 SiteEUIWN(kBtu/sf)  float64     6700     16   0.24
31              SiteEnergyUseWN(kBtu)  float64     6700     16   0.24
28                   SiteEUI(kBtu/sf)  float64     6699     17   0.25
32                 SourceEUI(kBtu/sf)  float64     6697     19   0.28
8        GHGEmissions(MetricTonsCO2e)  float64     6697     19   0.28
7                    Electricity(kWh)  float64     6697     19   0.28
6                   Electricity(kBtu)  float64     6697     19   0.28
15                   NaturalGas(kBtu)  float64     6697     19   0.28
16                 NaturalGas(therms)  float64     6697     19   0.28
34                     SteamUse(kBtu)  float64     6697     19   0.28
33               SourceEUIWN(kBtu/sf)  float64     6697     19   0.28
9   GHGEmissionsIntensity(kgCO2e/ft2)  float64     6697     19   0.28
13          ListOfAllPropertyUseTypes   object     6580    136   2.03
11          LargestPropertyUseTypeGFA  float64     6560    156   2.32
10             LargestPropertyUseType   object     6560    156   2.32
5                     ENERGYSTARScore  float64     5093   1623  24.17
27    SecondLargestPropertyUseTypeGFA  float64     3238   3478  51.79
26       SecondLargestPropertyUseType   object     3238   3478  51.79

In [23]:
pd.set_option("display.max_rows", None) # show full of showing rows
pd.set_option("display.max_columns", None) # show full of showing cols
pd.set_option("max_colwidth", None) # show full width of showing cols
data.head(5)

BuildingType ComplianceStatus  CouncilDistrictCode  DataYear  \
0  NonResidential        Compliant                    7      2015   
1  NonResidential        Compliant                    7      2015   
2  NonResidential        Compliant                    7      2015   
3  NonResidential        Compliant                    7      2015   
4  NonResidential        Compliant                    7      2015   

   DefaultData  ENERGYSTARScore  Electricity(kBtu)  Electricity(kWh)  \
0         True             65.0          3686160.0         1080307.0   
1         True             51.0          3905411.0         1144563.0   
2         True             18.0         49762435.0        14583930.0   
3         True              1.0          2769023.0          811521.0   
4         True             67.0          6066245.0         1777841.0   

   GHGEmissions(MetricTonsCO2e)  GHGEmissionsIntensity(kgCO2e/ft2)  \
0                        249.43                               2.64   
1                        263.51                               2.38   
2                       2061.48                               1.92   
3                       1936.34                              31.38   
4                        507.70                               4.02   

  LargestPropertyUseType  LargestPropertyUseTypeGFA   Latitude  \
0                  Hotel                    88434.0  47.612190   
1                  Hotel                    83880.0  47.613106   
2                  Hotel                   757243.0  47.613349   
3                  Hotel                    61320.0  47.614216   
4                  Hotel                   123445.0  47.613754   

       ListOfAllPropertyUseTypes   Longitude  NaturalGas(kBtu)  \
0                          Hotel -122.337997         1272388.0   
1     Hotel, Parking, Restaurant -122.333358         4448985.0   
2  Hotel, Parking, Swimming Pool -122.337699         3709900.0   
3                          Hotel -122.336609         2001894.0   
4  Hotel, Parking, Swimming Pool -122.340924         8763105.0   

   NaturalGas(therms) Neighborhood  NumberofBuildings  NumberofFloors  \
0             12724.0     DOWNTOWN                  1              12   
1             44490.0     DOWNTOWN                  1              11   
2             37099.0     DOWNTOWN                  1              41   
3             20019.0     DOWNTOWN                  1              10   
4             87631.0     DOWNTOWN                  1              18   

   OSEBuildingID PrimaryPropertyType  PropertyGFABuilding(s)  \
0              1               Hotel                   88434   
1              2               Hotel                   88502   
2              3               Hotel                  961990   
3              5               Hotel                   61320   
4              8               Hotel                  107430   

   PropertyGFAParking  PropertyGFATotal           PropertyName  \
0                   0             88434   MAYFLOWER PARK HOTEL   
1               15064            103566        PARAMOUNT HOTEL   
2                   0            961990           WESTIN HOTEL   
3                   0             61320              HOTEL MAX   
4               12460            119890  WARWICK SEATTLE HOTEL   

  SecondLargestPropertyUseType  SecondLargestPropertyUseTypeGFA  \
0                          NaN                              NaN   
1                      Parking                          15064.0   
2                      Parking                         100000.0   
3                          NaN                              NaN   
4                      Parking                          68009.0   

   SiteEUI(kBtu/sf)  SiteEUIWN(kBtu/sf)  SiteEnergyUse(kBtu)  \
0              78.9                80.3            6981428.0   
1              94.4                99.0            8354235.0   
2              96.6                99.7           73130656.0   
3             460.4               462.5           28229320.0  

<div style="background-color: #6D83C5;" >
    <h4 style="margin: auto; padding: 20px; color:#fff; ">3.2. Identifying and formatting categorical columns</h4>
</div>

In [24]:
data[list(set(data.columns) - set(data._get_numeric_data()))].head(10)

LargestPropertyUseType          BuildingType  \
0                                  Hotel        NonResidential   
1                                  Hotel        NonResidential   
2                                  Hotel        NonResidential   
3                                  Hotel        NonResidential   
4                                  Hotel        NonResidential   
5                         Police Station    Nonresidential COS   
6                                  Hotel        NonResidential   
7  Other - Entertainment/Public Assembly        NonResidential   
8                                  Hotel        NonResidential   
9                    Multifamily Housing  Multifamily MR (5-9)   

  TaxParcelIdentificationNumber                                 PropertyName  \
0                     659000030                         MAYFLOWER PARK HOTEL   
1                     659000220                              PARAMOUNT HOTEL   
2                     659000475                                 WESTIN HOTEL   
3                     659000640                                    HOTEL MAX   
4                     659000970                        WARWICK SEATTLE HOTEL   
5                     660000560               WEST PRECINCT (SEATTLE POLICE)   
6                     660000825                       CAMLIN WORLDMARK HOTEL   
7                     660000955                            PARAMOUNT THEATER   
8                     939000080  COURTYARD BY MARRIOTT - ALASKA BLDG SEATTLE   
9                     939000105                                LYON BUILDING   

    PrimaryPropertyType Neighborhood ComplianceStatus  \
0                 Hotel     DOWNTOWN        Compliant   
1                 Hotel     DOWNTOWN        Compliant   
2                 Hotel     DOWNTOWN        Compliant   
3                 Hotel     DOWNTOWN        Compliant   
4                 Hotel     DOWNTOWN        Compliant   
5                 Other     DOWNTOWN        Compliant   
6                 Hotel     DOWNTOWN        Compliant   
7                 Other     DOWNTOWN        Compliant   
8                 Hotel     DOWNTOWN        Compliant   
9  Mid-Rise Multifamily     DOWNTOWN        Compliant   

  SecondLargestPropertyUseType              ListOfAllPropertyUseTypes  
0                          NaN                                  Hotel  
1                      Parking             Hotel, Parking, Restaurant  
2                      Parking          Hotel, Parking, Swimming Pool  
3                          NaN                                  Hotel  
4                      Parking          Hotel, Parking, Swimming Pool  
5                          NaN                         Police Station  
6                          NaN                                  Hotel  
7                          NaN  Other - Entertainment/Public Assembly  
8                          NaN                                  Hotel  
9                          NaN                    Multifamily Housing

In [25]:
for col in list(set(data.columns) - set(data._get_numeric_data())):
    print(col)
    display(data[col].unique())

LargestPropertyUseType


array(['Hotel', 'Police Station', 'Other - Entertainment/Public Assembly',
       'Multifamily Housing', nan, 'Library',
       'Fitness Center/Health Club/Gym', 'Social/Meeting Hall',
       'Courthouse', 'Other', 'K-12 School', 'College/University',
       'Automobile Dealership', 'Office', 'Self-Storage Facility',
       'Retail Store', 'Senior Care Community', 'Medical Office',
       'Hospital (General Medical & Surgical)', 'Museum',
       'Repair Services (Vehicle, Shoe, Locksmith, etc)',
       'Other - Lodging/Residential', 'Residence Hall/Dormitory',
       'Other/Specialty Hospital', 'Financial Office',
       'Non-Refrigerated Warehouse', 'Distribution Center', 'Parking',
       'Worship Facility', 'Laboratory', 'Supermarket/Grocery Store',
       'Convention Center', 'Urgent Care/Clinic/Other Outpatient',
       'Other - Services', 'Strip Mall', 'Wholesale Club/Supercenter',
       'Refrigerated Warehouse', 'Other - Recreation', 'Lifestyle Center',
       'Other - Public S

BuildingType


array(['NonResidential', 'Nonresidential COS', 'Multifamily MR (5-9)',
       'SPS-District K-12', 'Multifamily LR (1-4)', 'Campus',
       'Multifamily HR (10+)', 'Nonresidential WA'], dtype=object)

TaxParcelIdentificationNumber


array(['659000030', '659000220', '659000475', ..., '1624049080',
       '7883603155', '7857002030'], dtype=object)

PropertyName


array(['MAYFLOWER PARK HOTEL', 'PARAMOUNT HOTEL', 'WESTIN HOTEL', ...,
       'Queen Anne Pool', 'South Park Community Center',
       'Van Asselt Community Center'], dtype=object)

PrimaryPropertyType


array(['Hotel', 'Other', 'Mid-Rise Multifamily', 'Mixed Use Property',
       'K-12 School', 'College/University', 'Small- and Mid-Sized Office',
       'Self-Storage Facility\n', 'Distribution Center', 'Large Office',
       'Retail Store', 'Low-Rise Multifamily', 'Senior Care Community',
       'Medical Office', 'Hospital', 'Residence Hall/Dormitory',
       'Non-Refrigerated Warehouse', 'Distribution Center\n',
       'SPS-District K-12', 'Worship Facility',
       'Supermarket/Grocery Store', 'Laboratory', 'Self-Storage Facility',
       'Refrigerated Warehouse', 'High-Rise Multifamily', 'Restaurant\n',
       'Restaurant', 'University', 'Warehouse', 'Residence Hall',
       'Supermarket / Grocery Store', 'Office'], dtype=object)

Neighborhood


array(['DOWNTOWN', 'SOUTHEAST', 'NORTHEAST', 'EAST', 'CENTRAL', 'NORTH',
       'MAGNOLIA / QUEEN ANNE', 'LAKE UNION', 'GREATER DUWAMISH',
       'BALLARD', 'NORTHWEST', 'SOUTHWEST', 'DELRIDGE', 'Central',
       'Ballard', 'North', 'Delridge', 'Northwest',
       'DELRIDGE NEIGHBORHOODS'], dtype=object)

ComplianceStatus


array(['Compliant', 'Not Compliant', 'Error - Correct Default Data',
       'Missing Data', 'Non-Compliant'], dtype=object)

SecondLargestPropertyUseType


array([nan, 'Parking', 'Office', 'Restaurant', 'K-12 School',
       'Non-Refrigerated Warehouse', 'Other - Education',
       'Vocational School', 'Retail Store',
       'Other - Entertainment/Public Assembly', 'Data Center',
       'Swimming Pool', 'Other', 'Courthouse', 'Residence Hall/Dormitory',
       'Distribution Center', 'Self-Storage Facility', 'Other - Services',
       'Laboratory', 'Movie Theater', 'Supermarket/Grocery Store',
       'Medical Office', 'Other - Recreation',
       'Convenience Store without Gas Station',
       'Fitness Center/Health Club/Gym', 'Performing Arts',
       'College/University', 'Other - Restaurant/Bar', 'Adult Education',
       'Multifamily Housing', 'Bar/Nightclub', 'Financial Office',
       'Repair Services (Vehicle, Shoe, Locksmith, etc)',
       'Other - Lodging/Residential', 'Bank Branch', 'Food Service',
       'Worship Facility', 'Senior Care Community',
       'Refrigerated Warehouse', 'Hotel', 'Food Sales',
       'Other/Specialty H

ListOfAllPropertyUseTypes


array(['Hotel', 'Hotel, Parking, Restaurant',
       'Hotel, Parking, Swimming Pool', 'Police Station',
       'Other - Entertainment/Public Assembly', 'Multifamily Housing',
       nan, 'Hotel, Parking', 'Data Center, Library, Parking',
       'Fitness Center/Health Club/Gym, Office, Swimming Pool',
       'Office, Other, Other - Lodging/Residential, Restaurant, Social/Meeting Hall',
       'Courthouse', 'Other', 'K-12 School', 'College/University',
       'Automobile Dealership',
       'Hotel, Parking, Retail Store, Swimming Pool',
       'K-12 School, Multifamily Housing, Office, Other - Entertainment/Public Assembly',
       'Self-Storage Facility', 'Non-Refrigerated Warehouse, Office',
       'Office, Other - Education',
       'Office, Self-Storage Facility, Vocational School',
       'Data Center, Office, Retail Store',
       'Non-Refrigerated Warehouse, Other, Retail Store',
       'Office, Retail Store',
       'Office, Other - Entertainment/Public Assembly, Restaurant',
   

<div class="alert alert-block alert-info">
    <b>Cleaning</b> data in the features
</div>

In [26]:
data["ComplianceStatus"].replace("Non-Compliant", "Not Compliant", inplace=True)
data["ComplianceStatus"].replace("Missing Data", np.nan, inplace=True)
data["ComplianceStatus"].replace("Error - Correct Default Data", np.nan, inplace=True)
data["Neighborhood"].replace("Ballard", "BALLARD", inplace=True)
data["Neighborhood"].replace("Central", "CENTRAL", inplace=True)
data["Neighborhood"].replace("North", "NORTH", inplace=True)
data["Neighborhood"].replace("Delridge", "DELRIDGE", inplace=True)
data["Neighborhood"].replace("DELRIDGE NEIGHBORHOODS", "DELRIDGE", inplace=True)
data["Neighborhood"].replace("Northwest", "NORTHWEST", inplace=True)

<div class="alert alert-block alert-info">
    <b>Changing</b> the type of data <b>Neighborhood</b>
</div>

In [27]:
for col in list(set(data.columns) - set(data._get_numeric_data())):
    data[col] = data[col].astype("category")

<div style="background-color: #6D83C5;" >
    <h4 style="margin: auto; padding: 20px; color:#fff; ">3.3. Filtering information</h4>
</div>

In [28]:
data["BuildingType"].unique()

['NonResidential', 'Nonresidential COS', 'Multifamily MR (5-9)', 'SPS-District K-12', 'Multifamily LR (1-4)', 'Campus', 'Multifamily HR (10+)', 'Nonresidential WA']
Categories (8, object): ['NonResidential', 'Nonresidential COS', 'Multifamily MR (5-9)', 'SPS-District K-12', 'Multifamily LR (1-4)', 'Campus', 'Multifamily HR (10+)', 'Nonresidential WA']

In [29]:
data = pd.DataFrame(data[~data["BuildingType"].str.contains("family", case=False, na=False).any(level=0)])

In [30]:
data["BuildingType"].unique()

['NonResidential', 'Nonresidential COS', 'SPS-District K-12', 'Campus', 'Nonresidential WA']
Categories (5, object): ['NonResidential', 'Nonresidential COS', 'SPS-District K-12', 'Campus', 'Nonresidential WA']

In [31]:
df_analysis(data, "data")


Analysis of data dataset
--------------------------------------------------------------------
- Dataset shape:                  3318 rows and 38 columns
- Total of NaN values:            4837
- Percentage of NaN:              3.84 %
- Total of full duplicates rows:  0
- Total of empty rows:            0
- Total of empty columns:         0
- Unique indexes:                 True

- Type object and records by columns   ( memory usage: 1.3 MB )
---------------------------------------------------------------------------


name      type  records  # NaN  % NaN
0                        BuildingType  category     3318      0   0.00
12                           Latitude   float64     3318      0   0.00
36                          YearBuilt     int64     3318      0   0.00
25                       PropertyName  category     3318      0   0.00
24                   PropertyGFATotal     int64     3318      0   0.00
23                 PropertyGFAParking     int64     3318      0   0.00
22             PropertyGFABuilding(s)     int64     3318      0   0.00
21                PrimaryPropertyType  category     3318      0   0.00
20                      OSEBuildingID     int64     3318      0   0.00
18                  NumberofBuildings     int64     3318      0   0.00
17                       Neighborhood  category     3318      0   0.00
14                          Longitude   float64     3318      0   0.00
19                     NumberofFloors     int64     3318      0   0.00
3                            DataYear     int64     3318      0   0.00
4                         DefaultData      bool     3318      0   0.00
2                 CouncilDistrictCode     int64     3318      0   0.00
37                            ZipCode     int64     3318      0   0.00
35      TaxParcelIdentificationNumber  category     3317      1   0.03
32                 SourceEUI(kBtu/sf)   float64     3309      9   0.27
15                   NaturalGas(kBtu)   float64     3309      9   0.27
8        GHGEmissions(MetricTonsCO2e)   float64     3309      9   0.27
7                    Electricity(kWh)   float64     3309      9   0.27
6                   Electricity(kBtu)   float64     3309      9   0.27
16                 NaturalGas(therms)   float64     3309      9   0.27
34                     SteamUse(kBtu)   float64     3309      9   0.27
33               SourceEUIWN(kBtu/sf)   float64     3309      9   0.27
9   GHGEmissionsIntensity(kgCO2e/ft2)   float64     3309      9   0.27
30                SiteEnergyUse(kBtu)   float64     3309      9   0.27
28                   SiteEUI(kBtu/sf)   float64     3308     10   0.30
29                 SiteEUIWN(kBtu/sf)   float64     3308     10   0.30
31              SiteEnergyUseWN(kBtu)   float64     3308     10   0.30
13          ListOfAllPropertyUseTypes  category     3255     63   1.90
10             LargestPropertyUseType  category     3247     71   2.14
11          LargestPropertyUseTypeGFA   float64     3247     71   2.14
1                    ComplianceStatus  category     3216    102   3.07
5                     ENERGYSTARScore   float64     2211   1107  33.36
26       SecondLargestPropertyUseType  category     1667   1651  49.76
27    SecondLargestPropertyUseTypeGFA   float64     1667   1651  49.76

<div class="alert alert-block alert-success">
 At this point, we get a dataset shape 3318 rows and 38
</div>

In [32]:
pd.set_option("display.max_rows", None) # show full of showing rows
pd.set_option("display.max_columns", None) # show full of showing cols
pd.set_option("max_colwidth", None) # show full width of showing cols
data.head(10)

BuildingType ComplianceStatus  CouncilDistrictCode  DataYear  \
0       NonResidential        Compliant                    7      2015   
1       NonResidential        Compliant                    7      2015   
2       NonResidential        Compliant                    7      2015   
3       NonResidential        Compliant                    7      2015   
4       NonResidential        Compliant                    7      2015   
5   Nonresidential COS        Compliant                    7      2015   
6       NonResidential        Compliant                    7      2015   
7       NonResidential        Compliant                    7      2015   
8       NonResidential        Compliant                    7      2015   
10      NonResidential        Compliant                    7      2015   

    DefaultData  ENERGYSTARScore  Electricity(kBtu)  Electricity(kWh)  \
0          True             65.0          3686160.0         1080307.0   
1          True             51.0          3905411.0         1144563.0   
2          True             18.0         49762435.0        14583930.0   
3          True              1.0          2769023.0          811521.0   
4          True             67.0          6066245.0         1777841.0   
5          True              NaN          7271004.0         2130921.0   
6          True             25.0          2679698.0          785342.0   
7          True              NaN          4108004.0         1203937.0   
8          True             46.0          7355649.0         2155728.0   
10         True             48.0          6337531.0         1857347.0   

    GHGEmissions(MetricTonsCO2e)  GHGEmissionsIntensity(kgCO2e/ft2)  \
0                         249.43                               2.64   
1                         263.51                               2.38   
2                        2061.48                               1.92   
3                        1936.34                              31.38   
4                         507.70                               4.02   
5                         304.62                               2.81   
6                         208.46                               2.37   
7                         199.99                               1.77   
8                         331.61                               1.83   
10                        576.63                               3.59   

                   LargestPropertyUseType  LargestPropertyUseTypeGFA  \
0                                   Hotel                    88434.0   
1                                   Hotel                    83880.0   
2                                   Hotel                   757243.0   
3                                   Hotel                    61320.0   
4                                   Hotel                   123445.0   
5                          Police Station                    88830.0   
6                                   Hotel                    81352.0   
7   Other - Entertainment/Public Assembly                   102761.0   
8                                   Hotel                   163984.0   
10                                    NaN                        NaN   

     Latitude              ListOfAllPropertyUseTypes   Longitude  \
0   47.612190                                  Hotel -122.337997   
1   47.613106             Hotel, Parking, Restaurant -122.333358   
2   47.613349          Hotel, Parking, Swimming Pool -122.337699   
3   47.614216                                  Hotel -122.336609   
4   47.613754          Hotel, Parking, Swimming Pool -122.340924   
5   47.616439                         Police Station -122.336764   
6   47.614114                                  Hotel -122.332741   
7   47.612902  Other - Entertainment/Public Assembly -122.331309   
8   47.602589                                  Hotel -122.332553   
10  47.607121                                    NaN -122.334319   

    NaturalGas(kBtu)  NaturalGas(therms) Neighborhood  NumberofBuil

<div style="background-color: #6D83C5;" >
    <h4 style="margin: auto; padding: 20px; color:#fff; ">3.4. Selecting the features</h4>
</div>

In [33]:
LIST_FEATURES = [
    "OSEBuildingID", "DataYear", "PrimaryPropertyType", "NumberofFloors", "NumberofBuildings", "YearBuilt", "ZipCode", "Neighborhood",
    "ENERGYSTARScore", "Electricity(kBtu)", "NaturalGas(kBtu)", "SiteEUI(kBtu/sf)", "SiteEUIWN(kBtu/sf)", "SiteEnergyUse(kBtu)", 
    "SiteEnergyUseWN(kBtu)", "SourceEUI(kBtu/sf)", "SourceEUIWN(kBtu/sf)", "SteamUse(kBtu)", "PropertyGFABuilding(s)", "PropertyGFAParking",
    "PropertyGFATotal"    
]

In [34]:
df_to_work = pd.DataFrame(data[LIST_FEATURES])

In [58]:
df_analysis(df_to_work, "df_to_work", type_analysis="complete")


Analysis of df_to_work dataset
--------------------------------------------------------------------
- Dataset shape:                  3318 rows and 21 columns
- Total of NaN values:            1191
- Percentage of NaN:              1.71 %
- Total of full duplicates rows:  0
- Total of empty rows:            0
- Total of empty columns:         0
- Unique indexes:                 True

- Type object and records by columns      ( memory usage: 655.9 KB )
--------------------------------------------------------------------


name      type  records  unique  # NaN  % NaN  \
0            OSEBuildingID     int64     3318    1698      0   0.00   
6                  ZipCode     int64     3318      78      0   0.00   
19      PropertyGFAParking     int64     3318     366      0   0.00   
18  PropertyGFABuilding(s)     int64     3318    1694      0   0.00   
1                 DataYear     int64     3318       2      0   0.00   
7             Neighborhood  category     3318      13      0   0.00   
20        PropertyGFATotal     int64     3318    1667      0   0.00   
5                YearBuilt     int64     3318     113      0   0.00   
4        NumberofBuildings     int64     3318      18      0   0.00   
3           NumberofFloors     int64     3318      45      0   0.00   
2      PrimaryPropertyType  category     3318      30      0   0.00   
9        Electricity(kBtu)   float64     3309    3296      9   0.27   
13     SiteEnergyUse(kBtu)   float64     3309    3291      9   0.27   
15      SourceEUI(kBtu/sf)   float64     3309    2464      9   0.27   
16    SourceEUIWN(kBtu/sf)   float64     3309    2475      9   0.27   
17          SteamUse(kBtu)   float64     3309     229      9   0.27   
10        NaturalGas(kBtu)   float64     3309    2362      9   0.27   
11        SiteEUI(kBtu/sf)   float64     3308    1824     10   0.30   
12      SiteEUIWN(kBtu/sf)   float64     3308    1843     10   0.30   
14   SiteEnergyUseWN(kBtu)   float64     3308    3272     10   0.30   
8          ENERGYSTARScore   float64     2211     100   1107  33.36   

            mean       min           25%           50%           75%  \
0   1.595182e+04       1.0  5.790000e+02  2.112200e+04  2.455675e+04   
6   5.818633e+04       0.0  1.839000e+04  1.958400e+04  9.810900e+04   
19  1.330330e+04      -2.0  0.000000e+00  0.000000e+00  0.000000e+00   
18  1.023639e+05  -50550.0  2.850775e+04  4.736800e+04  9.447150e+04   
1   2.015503e+03    2015.0  2.015000e+03  2.016000e+03  2.016000e+03   
7            NaN       NaN           NaN           NaN           NaN   
20  1.156672e+05   11285.0  2.951225e+04  4.924600e+04  1.047860e+05   
5   1.961537e+03    1900.0  1.930000e+03  1.965000e+03  1.989000e+03   
4   1.116637e+00       0.0  1.000000e+00  1.000000e+00  1.000000e+00   
3   4.116938e+00       0.0  1.000000e+00  2.000000e+00  4.000000e+00   
2            NaN       NaN           NaN           NaN           NaN   
9   5.636556e+06 -115417.0  7.236670e+05  1.623657e+06  4.878886e+06   
13  8.068625e+06       0.0  1.223425e+06  2.505271e+06  6.917280e+06   
15  1.754400e+02      -2.0  7.620000e+01  1.313000e+02  2.049000e+02   
16  1.786557e+02      -2.1  8.080000e+01  1.348000e+02  2.078000e+02   
17  4.934034e+05       0.0  0.000000e+00  0.000000e+00  0.000000e+00   
10  1.876485e+06       0.0  0.000000e+00  4.797690e+05  1.471375e+06   
11  7.160834e+01       0.0  3.320000e+01  5.170000e+01  8.170000e+01   
12  7.483298e+01       0.0  3.590000e+01  5.480000e+01  8.612500e+01   
14  7.991069e+06       0.0  1.327599e+06  2.686911e+06  7.175216e+06   
8   6.481592e+01       1.0  4.700000e+01  7.200000e+01  8.900000e+01   

             max           std  
0   5.022600e+04  1.341190e+04  
6   9.819900e+04  3.979171e+04  
19  5.126080e+05  4.359662e+04  
18  9.320156e+06  2.340749e+05  
1   2.016000e+03  5.000680e-01  
7            NaN           NaN  
20  9.320156e+06  2.511222e+05  
5   2.015000e+03  3.270977e+01  
4   1.110000e+02  2.219345e+00  
3   9.900000e+01  6.562531e+00  
2            NaN           NaN  
9   6.570744e+08  1.740900e+07  
13  8.739237e+08  2.525462e+07  
15  2.620000e+03  1.807871e+02  
16  2.620000e+03  1.806337e+02  
17  1.349435e+08  5.226678e+06  
10  2.979090e+08  7.898387e+06  
11  8.344000e+02  7.237218e+01  
12  8.344000e+02  7.382939e+01  
14  4.716139e+08  2.068644e+07  
8   1.000000e+02  2.854989e+01

In [36]:
df_to_work.head(10)

OSEBuildingID  DataYear PrimaryPropertyType  NumberofFloors  \
0               1      2015               Hotel              12   
1               2      2015               Hotel              11   
2               3      2015               Hotel              41   
3               5      2015               Hotel              10   
4               8      2015               Hotel              18   
5               9      2015               Other               2   
6              10      2015               Hotel              11   
7              11      2015               Other               8   
8              12      2015               Hotel              15   
10             15      2015               Hotel              11   

    NumberofBuildings  YearBuilt  ZipCode Neighborhood  ENERGYSTARScore  \
0                   1       1927    18081     DOWNTOWN             65.0   
1                   1       1996    18081     DOWNTOWN             51.0   
2                   1       1969    18081     DOWNTOWN             18.0   
3                   1       1926    18081     DOWNTOWN              1.0   
4                   1       1980    19576     DOWNTOWN             67.0   
5                   1       1999    19576     DOWNTOWN              NaN   
6                   1       1926    18081     DOWNTOWN             25.0   
7                   1       1926    18081     DOWNTOWN              NaN   
8                   1       1904    18379     DOWNTOWN             46.0   
10                  1       1969    18081     DOWNTOWN             48.0   

    Electricity(kBtu)  ...  SiteEUI(kBtu/sf)  SiteEUIWN(kBtu/sf)  \
0           3686160.0  ...              78.9                80.3   
1           3905411.0  ...              94.4                99.0   
2          49762435.0  ...              96.6                99.7   
3           2769023.0  ...             460.4               462.5   
4           6066245.0  ...             120.1               122.1   
5           7271004.0  ...             135.7               146.9   
6           2679698.0  ...              76.9                79.6   
7           4108004.0  ...              62.5                71.8   
8           7355649.0  ...              77.0                82.9   
10          6337531.0  ...             101.1               107.2   

    SiteEnergyUse(kBtu)  SiteEnergyUseWN(kBtu)  SourceEUI(kBtu/sf)  \
0             6981428.0              7097539.0               173.5   
1             8354235.0              8765788.0               191.3   
2            73130656.0             75506272.0               242.7   
3            28229320.0             28363444.0               636.3   
4            14829099.0             15078243.0               228.8   
5            12051984.0             13045258.0               313.5   
6             6252842.0              6477493.0               149.5   
7             6426022.0              7380086.0               152.2   
8            12633744.0             13589025.0               174.6   
10           14719853.0             15615929.0               200.8   

    SourceEUIWN(kBtu/sf)  SteamUse(kBtu)  PropertyGFABuilding(s)  \
0                  175.1       2023032.0                   88434   
1                  195.2             0.0                   88502   
2                  246.5      19660404.0                  961990   
3                  643.2      23458518.0                   61320   
4                  227.1             0.0                  107430   
5                  321.6             0.0                   60090   
6                  158.2             0.0                   83008   
7                  160.4       2003108.0                  102761   
8                  186.8             0.0                  163984   
10                 205.2       3623542.0                  133884   

    PropertyGFAParking  PropertyGFATotal  
0                    0             88434  
1                15064            103566  
2                    0            961990  

<div style="background-color: #506AB9;" >
    <h3 style="margin: auto; padding: 20px; color:#fff; ">4. Treating Outliers</h3>
</div>

<div style="background-color: #6D83C5;" >
    <h4 style="margin: auto; padding: 20px; color:#fff; ">4.1. Treating Outliers with Z Score</h4>
</div>

<div class="alert alert-block alert-info">
    Deleting <b>outliers</b> with <b>Z Score</b>
</div>

In [61]:
df_without_outliers_by_zscore = pd.DataFrame()

for property_type in df_to_work["PrimaryPropertyType"].unique():
    
    # for each property type, we are going to define the max and min threshold to replace outliers
    df_temp_outliers = df_to_work[df_to_work["PrimaryPropertyType"] == property_type].copy()
    
    for col in set(df_temp_outliers._get_numeric_data()):
        
        # Defining the Z Score
        df_temp_outliers[col + "_zscore"] = (df_temp_outliers[col] - df_temp_outliers[col].mean()) / df_temp_outliers[col].std()
        df_temp_outliers.loc[(df_temp_outliers[col + "_zscore"] > 3) | (df_temp_outliers[col + "_zscore"] < -3), col] = np.nan
        
        del df_temp_outliers[col + "_zscore"]
        
    df_without_outliers_by_zscore = pd.concat([df_without_outliers_by_zscore, df_temp_outliers])

In [62]:
df_analysis(df_without_outliers_by_zscore, "df_without_outliers_by_zscore", type_analysis="complete")


Analysis of df_without_outliers_by_zscore dataset
--------------------------------------------------------------------
- Dataset shape:                  3318 rows and 21 columns
- Total of NaN values:            2102
- Percentage of NaN:              3.02 %
- Total of full duplicates rows:  0
- Total of empty rows:            0
- Total of empty columns:         0
- Unique indexes:                 True

- Type object and records by columns      ( memory usage: 526.9 KB )
--------------------------------------------------------------------


name      type  records  unique  # NaN  % NaN  \
2      PrimaryPropertyType  category     3318      30      0   0.00   
5                YearBuilt   float64     3318     113      0   0.00   
7             Neighborhood  category     3318      13      0   0.00   
1                 DataYear   float64     3313       2      5   0.15   
6                  ZipCode   float64     3307      77     11   0.33   
0            OSEBuildingID   float64     3291    1682     27   0.81   
4        NumberofBuildings   float64     3263      13     55   1.66   
3           NumberofFloors   float64     3258      39     60   1.81   
17          SteamUse(kBtu)   float64     3255     175     63   1.90   
18  PropertyGFABuilding(s)   float64     3254    1660     64   1.93   
20        PropertyGFATotal   float64     3253    1635     65   1.96   
13     SiteEnergyUse(kBtu)   float64     3252    3234     66   1.99   
12      SiteEUIWN(kBtu/sf)   float64     3250    1792     68   2.05   
14   SiteEnergyUseWN(kBtu)   float64     3250    3214     68   2.05   
16    SourceEUIWN(kBtu/sf)   float64     3249    2417     69   2.08   
11        SiteEUI(kBtu/sf)   float64     3248    1769     70   2.11   
15      SourceEUI(kBtu/sf)   float64     3248    2408     70   2.11   
19      PropertyGFAParking   float64     3248     330     70   2.11   
9        Electricity(kBtu)   float64     3245    3232     73   2.20   
10        NaturalGas(kBtu)   float64     3244    2297     74   2.23   
8          ENERGYSTARScore   float64     2194     100   1124  33.88   

            mean       min           25%         50%           75%  \
2            NaN       NaN           NaN         NaN           NaN   
5   1.961537e+03    1900.0  1.930000e+03     1965.00  1.989000e+03   
7            NaN       NaN           NaN         NaN           NaN   
1   2.015503e+03    2015.0  2.015000e+03     2016.00  2.016000e+03   
6   5.817348e+04       0.0  1.839000e+04    19584.00  9.810900e+04   
0   1.573477e+04       1.0  5.775000e+02    21108.00  2.452250e+04   
4   1.038308e+00       0.0  1.000000e+00        1.00  1.000000e+00   
3   3.742173e+00       0.0  1.000000e+00        2.00  4.000000e+00   
17  3.261433e+05       0.0  0.000000e+00        0.00  0.000000e+00   
18  8.630355e+04  -50550.0  2.830400e+04    46261.50  8.987500e+04   
20  9.741674e+04   11285.0  2.925800e+04    48117.00  9.944000e+04   
13  6.912324e+06       0.0  1.207988e+06  2454836.50  6.618646e+06   
12  7.062378e+01       0.0  3.540000e+01       54.15  8.357500e+01   
14  7.023268e+06       0.0  1.302417e+06  2624264.50  6.867424e+06   
16  1.669700e+02      -2.1  7.980000e+01      132.90  2.027000e+02   
11  6.727777e+01       0.0  3.287500e+01       51.00  7.940000e+01   
15  1.633946e+02      -2.0  7.530000e+01      129.05  1.991250e+02   
19  9.953457e+03      -2.0  0.000000e+00        0.00  0.000000e+00   
9   4.745226e+06 -115417.0  7.154750e+05  1581352.00  4.647474e+06   
10  1.492391e+06       0.0  0.000000e+00   445910.00  1.370250e+06   
8   6.527712e+01       1.0  4.800000e+01       73.00  8.900000e+01   

             max           std  
2            NaN           NaN  
5   2.015000e+03  3.270977e+01  
7            NaN           NaN  
1   2.016000e+03  5.000634e-01  
6   9.819900e+04  3.979182e+04  
0   5.022600e+04  1.315410e+04  
4   2.700000e+01  6.640020e-01  
3   4.900000e+01  5.068624e+00  
17  1.349435e+08  4.790492e+06  
18  2.200000e+06  1.266875e+05  
20  2.200000e+06  1.426816e+05  
13  2.958126e+08  1.627684e+07  
12  6.931000e+02  6.231692e+01  
14  2.977410e+08  1.635781e+07  
16  1.195100e+03  1.394715e+02  
11  6.967000e+02  6.038216e+01  
15  1.206700e+03  1.383581e+02  
19  3.037070e+05  3.138673e+04  
9   1.708916e+08  9.920529e+06  
10  1.381912e+08  5.002848e+06  
8   1.000000e+02  2.816708e+01

<div style="background-color: #6D83C5;" >
    <h4 style="margin: auto; padding: 20px; color:#fff; ">4.2. Comparison of Outliers treatment</h4>
</div>

<div class="alert alert-block alert-info">
    Identifying the <b>PrimaryPropertyType</b> with more records to plot the comparison with <b>Z Score result</b>
</div>

In [40]:
df_to_work.groupby("PrimaryPropertyType")["OSEBuildingID"].count().sort_values(ascending=False)

PrimaryPropertyType
Small- and Mid-Sized Office    588
Other                          502
Large Office                   344
K-12 School                    275
Mixed Use Property             220
Retail Store                   191
Warehouse                      187
Non-Refrigerated Warehouse     187
Hotel                          149
Worship Facility               143
Medical Office                  82
Distribution Center             55
Distribution Center\n           51
Supermarket / Grocery Store     40
Senior Care Community           39
Supermarket/Grocery Store       36
Self-Storage Facility           29
Self-Storage Facility\n         27
Refrigerated Warehouse          25
University                      24
Residence Hall                  21
College/University              21
Hospital                        20
Residence Hall/Dormitory        15
Restaurant                      13
Restaurant\n                    11
Laboratory                      11
Low-Rise Multifamily             5


<div class="alert alert-block alert-info">
    <b>Plotting</b> the comparison between data before and after treating <b>outliers</b> with <b>Z Score</b>
</div>

In [ ]:
LIST_FEATURES = ["ENERGYSTARScore", "SourceEUI(kBtu/sf)", "NaturalGas(kBtu)"]
property_type = "Large Office"

df_temp = df_to_work[df_to_work["PrimaryPropertyType"] == property_type].copy()
df_temp_by_zscore = df_without_outliers_by_zscore[df_without_outliers_by_zscore["PrimaryPropertyType"] == property_type].copy()

for col in LIST_FEATURES:

    # Creating a figure composed of two matplotlib.Axes objects (ax_box and ax_hist)
    fig, ((ax_box, ax_box_zscore), (ax_hist, ax_hist_zscore)) = \
                                                        plt.subplots(nrows=2, ncols=2, figsize=[14,8], sharex=True, gridspec_kw={"height_ratios": (.15, .85)})

    # Title for the complete RR078466118RUfigure
    fig.suptitle((col.capitalize() + " in " + property_type), fontsize="x-large", fontweight="bold", size=20)

    # Assigning a graph to each ax
    ax_box.title.set_text("Before treating outliers")
    ax_box_zscore.title.set_text("After zscore")

    sns.boxplot(data=df_temp, x=col, ax=ax_box)
    sns.boxplot(data=df_temp_by_zscore, x=col, ax=ax_box_zscore)

    sns.histplot(data=df_temp, x=col, bins=100, binwidth=3, kde=True, ax=ax_hist)
    sns.histplot(data=df_temp_by_zscore, x=col, bins=100, binwidth=3, kde=True, ax=ax_hist_zscore)

    # Remove x axis name for the boxplot
    ax_box.set(xlabel="")
    ax_box.set(yticks=[])
    ax_box_zscore.set(xlabel="")
    ax_box_zscore.set(yticks=[])

    plt.tight_layout()
    plt.savefig("img/outliers.png")
    sns.despine(fig)

In [42]:
df_analysis(df_without_outliers_by_zscore, "df_without_outliers_by_zscore")


Analysis of df_without_outliers_by_zscore dataset
--------------------------------------------------------------------
- Dataset shape:                  3318 rows and 21 columns
- Total of NaN values:            2102
- Percentage of NaN:              3.02 %
- Total of full duplicates rows:  0
- Total of empty rows:            0
- Total of empty columns:         0
- Unique indexes:                 True

- Type object and records by columns   ( memory usage: 526.9 KB )
---------------------------------------------------------------------------


name      type  records  # NaN  % NaN
2      PrimaryPropertyType  category     3318      0   0.00
5                YearBuilt   float64     3318      0   0.00
7             Neighborhood  category     3318      0   0.00
1                 DataYear   float64     3313      5   0.15
6                  ZipCode   float64     3307     11   0.33
0            OSEBuildingID   float64     3291     27   0.81
4        NumberofBuildings   float64     3263     55   1.66
3           NumberofFloors   float64     3258     60   1.81
17          SteamUse(kBtu)   float64     3255     63   1.90
18  PropertyGFABuilding(s)   float64     3254     64   1.93
20        PropertyGFATotal   float64     3253     65   1.96
13     SiteEnergyUse(kBtu)   float64     3252     66   1.99
12      SiteEUIWN(kBtu/sf)   float64     3250     68   2.05
14   SiteEnergyUseWN(kBtu)   float64     3250     68   2.05
16    SourceEUIWN(kBtu/sf)   float64     3249     69   2.08
11        SiteEUI(kBtu/sf)   float64     3248     70   2.11
15      SourceEUI(kBtu/sf)   float64     3248     70   2.11
19      PropertyGFAParking   float64     3248     70   2.11
9        Electricity(kBtu)   float64     3245     73   2.20
10        NaturalGas(kBtu)   float64     3244     74   2.23
8          ENERGYSTARScore   float64     2194   1124  33.88